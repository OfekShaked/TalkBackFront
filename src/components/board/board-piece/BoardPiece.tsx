import React, { useState, useEffect } from 'react';
import { handleError } from '../../../services/errorHandling.service'
interface BoardPieceProps {
    color: String;
    position: number;
    isDead: Boolean;
    row: number;
    onClick: any | null;
    isOptional: Boolean;
    children?: JSX.Element;
    keyValue: any;
}

const BoardPiece = (props: BoardPieceProps) => {
    const { color, position, isDead, row, isOptional, keyValue } = props;
    //position can be 0-25 where 0 and 25 is outside(eaten)
    const [topOrBottom, setTopOrBottom] = useState<String>("");

    const [column, setColumn] = useState<number>();

    useEffect(() => {
        handleColumn();
    }, [position])


    const handleColumn = (): void => {
        try {
            setTopOrBottom(position > 12 ? "top" : "bottom");
            if (position > 12) {
                let columnNum: number;
                switch (position) {
                    case 13: columnNum = 12; break;
                    case 14: columnNum = 11; break;
                    case 15: columnNum = 10; break;
                    case 16: columnNum = 9; break;
                    case 17: columnNum = 8; break;
                    case 18: columnNum = 7; break;
                    case 19: columnNum = 6; break;
                    case 20: columnNum = 5; break;
                    case 21: columnNum = 4; break;
                    case 22: columnNum = 3; break;
                    case 23: columnNum = 2; break;
                    case 24: columnNum = 1; break;
                    default: columnNum = 0; break;

                }
                if (position <= 0) {
                    columnNum = 0;
                }
                setColumn(columnNum);
            }
            else {
                setColumn(position);
            }
        } catch (error) {
            handleError(error)
        }
    }

    return (
        <React.Fragment key={keyValue}>
            {!isDead ?
                <div className={`piece_${color} ${topOrBottom}_row_${row} column_${column} ${isOptional ? "optional" : ""} ${props.onClick ? "canBeMoved" : ""}`} draggable="true" onClick={props.onClick}>
                    {props.children}
                </div>
                : <></>}
        </React.Fragment>
    )
}
export default BoardPiece;