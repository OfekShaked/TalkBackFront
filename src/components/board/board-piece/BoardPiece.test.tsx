import React from 'react';
import renderer from 'react-test-renderer';
import BoardPiece from './BoardPiece';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import { mount, configure } from 'enzyme';

configure({ adapter: new Adapter() });

const onClick = jest.fn();

describe('Board piece tests', () => {
    it('renders BoardPiece component', async () => {
        const component = renderer.create(
                <BoardPiece color={"white"} position={21} isDead={false} row={1} onClick={onClick} isOptional={false} keyValue={1}></BoardPiece>
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('displays white color piece', async () => {
        const component = mount(
            <BoardPiece color={"white"} position={21} isDead={false} row={1} onClick={onClick} isOptional={false} keyValue={1}></BoardPiece>
            );
        const piece = component.find('div');
        expect(piece.props().className).toContain("white");
    });

    it('displays black color piece', async () => {
        const component = mount(
            <BoardPiece color={"black"} position={21} isDead={false} row={1} onClick={onClick} isOptional={false} keyValue={1}></BoardPiece>
            );
        const piece = component.find('div');
        expect(piece.props().className).toContain("black");
    });
    
    it('correct column for piece position over 12', async () => {
        const component = mount(
            <BoardPiece color={"white"} position={21} isDead={false} row={1} onClick={onClick} isOptional={false} keyValue={1}></BoardPiece>
            );
        const piece = component.find('div');
        expect(piece.props().className).toContain("column_4");
    });
    
    it('correct column for piece position under 12', async () => {
        const component = mount(
            <BoardPiece color={"white"} position={11} isDead={false} row={1} onClick={onClick} isOptional={false} keyValue={1}></BoardPiece>
            );
        const piece = component.find('div');
        expect(piece.props().className).toContain("column_11");
    });

    it('clicking on piece is working', async () => {
        const component = mount(
            <BoardPiece color={"white"} position={11} isDead={false} row={1} onClick={onClick} isOptional={false} keyValue={1}></BoardPiece>
            );
        const piece = component.find('div');
        piece.simulate("click")
        expect(onClick).toBeCalled();
    });

    it('dead piece should not display', async () => {
        const component = mount(
            <BoardPiece color={"white"} position={11} isDead={true} row={1} onClick={onClick} isOptional={false} keyValue={1}></BoardPiece>
            );
        const piece = component.find('div');
        expect(piece.length).toBe(0);
    });

    it('optional piece should display', async () => {
        const component = mount(
            <BoardPiece color={"white"} position={11} isDead={false} row={1} onClick={onClick} isOptional={true} keyValue={1}></BoardPiece>
            );
        const piece = component.find('div');
        expect(piece.props().className).toContain("optional");
    });
})

