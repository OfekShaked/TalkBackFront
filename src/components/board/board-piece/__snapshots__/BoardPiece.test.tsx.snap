// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Board piece tests renders BoardPiece component 1`] = `
<div
  className="piece_white _row_1 column_undefined  canBeMoved"
  draggable="true"
  onClick={[MockFunction]}
/>
`;
